### PR TITLE
avoid copies in matrix_cl & add OpenCL CPU tests

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary
Done by @t4c1 : 

When a `matrix_cl` is created from data on host, it copies the data to OpenCL device. This is not necessary if the OpenCL device in use uses same memory as host (when the device is CPU or integrated GPU). 

This PR changes `matrix_cl` to instead use the same data if the device uses same memory as host. This is limited to when `matrix_cl` is constructed from a non-const value. Copying from `matrix_cl` to host (`from_matrix_cl`) is not optimized yet, as it would require returning of `Eigen::Map`.

Also added `pinned_matrix` class, which works as an `Eigen::Matrix`, except using OpenCL implementation allocated (likely pinned) memory. When making `matrix_cl` from `pinned_matrix` we should get no-copy behavior on CPU/integrated GPU and a bit faster copying on discrete GPU. `pinned_matrix` is also returned from `ref_type_for_opencl`.

My parts:

This PR will add OpenCL CPU tests. These tests are short and with #2182 making OpenCL on the CPU more efficient and useful, we should test this. Especially given it's a very short test.

Right now, this PR builds off of #2182.

## Tests

/

## Side Effects

/

## Release notes

`matrix_cl` now uses host-allocated memory to avoid copies if the OpenCL device uses the same memory as host. Added `pinned_matrix` class.

## Checklist

- [x] Copyright holder: Rok Češnovar, Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Thanks @serban-nicusor-toptal for setting up the OpenCL CPU!

There is an issue with test/unit/math/opencl/kernel_generator/block_test.cpp on the CPU that seems to be a legit one. @t4c1 could you look at that when you have the chance. I guess we can add those fixes either here on in #2182, either way, is fine by me.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal hey, would it be possible to also set up the OpenCL CPU runtime on the gelman-group-linux machine? It should be simple to install. We dont want these tests relying on a single machine as that becomes a huge bottleneck if it goes down for some reason.

**Comment by serban-nicusor-toptal** [comment]
Sure @rok-cesnovar I'm installing it now.

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar I've installed the [runtime](https://github.com/intel/compute-runtime/blob/master/opencl/doc/DISTRIBUTIONS.md).
Here you can find all the libOpenCL for both nvidia and intel:

```
nvidia-libopencl1-340: /usr/lib/i386-linux-gnu/libOpenCL.so.1
nvidia-libopencl1-340: /usr/lib/i386-linux-gnu/libOpenCL.so.1.0
nvidia-libopencl1-340: /usr/lib/i386-linux-gnu/libOpenCL.so.1.0.0
nvidia-libopencl1-340: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1
nvidia-libopencl1-340: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0
nvidia-libopencl1-340: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0
ocl-icd-libopencl1: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1
ocl-icd-libopencl1: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1.0.0
ocl-icd-libopencl1: /usr/share/man/man7/libOpenCL.so.7.gz
ocl-icd-opencl-dev: /usr/lib/x86_64-linux-gnu/libOpenCL.so
```

It doesn't show up in `clinfo` so if it's not working properly please let me know to look into it.

**Comment by rok-cesnovar** [comment]
Hey @serban-nicusor-toptal, this runtime is for Intel GPUs and latest CPUs I believe. 

You need to install Intel® CPU Runtime for OpenCL™ Applications 18.1 that can be found here https://software.intel.com/content/www/us/en/develop/articles/opencl-drivers.html
You do need to register. If you dont want to register I can send you the tarball. Thanks!

**Comment by serban-nicusor-toptal** [comment]
My bad, sorry @rok-cesnovar
I've installed the proper one now, it shows up in `clinfo`. Is there anything else I can help with ?

**Comment by rok-cesnovar** [comment]
Thanks Nic!

**Comment by rok-cesnovar** [comment]
This ended up being “my” PR but my changes are only in Jenkinsfile.

@SteveBronder approved the changes in #2182 and I approve the additional bugfixes done for this PR. @t4c1 if you approve the changes in Jenkinsfile then go ahead and formally approve this.

File: Jenkinsfile [hunk 1]
```diff
@@ -224,25 +224,87 @@ pipeline {
...
-                        runTests("test/unit/math/prim/fun/mdivide_left_tri_test.cpp")
-                        runTests("test/unit/math/prim/fun/mdivide_right_tri_test.cpp")
-                        runTests("test/unit/math/prim/fun/multiply_test.cpp")
-                        runTests("test/unit/math/rev/fun/mdivide_left_tri_test.cpp")
-                        runTests("test/unit/math/rev/fun/multiply_test.cpp")
+                        script {
+                            if (isUnix()) {
+                                deleteDir()
+                                unstash 'MathSetup'
+                                sh "clinfo"
```
Comment (t4c1) [inline Jenkinsfile:234]: Do we need clinfo?

File: Jenkinsfile [hunk 2]
```diff
@@ -224,25 +224,87 @@ pipeline {
...
+                                bat 'echo LDFLAGS_OPENCL= -L"C:\\Program Files (x86)\\IntelSWTools\\system_studio_2020\\OpenCL\\sdk\\lib\\x64" -lOpenCL >> make/local'
+                                bat "mingw32-make.exe -f make/standalone math-libs"
+                                runTestsWin("test/unit/math/prim/fun/multiply_test.cpp", false, false)
+                                runTestsWin("test/unit/math/opencl", false, false)
+                                runTestsWin("test/unit/multiple_translation_units_test.cpp", false, false)
+                                runTestsWin("test/unit/math/prim/fun/gp_exp_quad_cov_test.cpp", false, false)
+                                runTestsWin("test/unit/math/prim/fun/mdivide_left_tri_test.cpp", false, false)
+                                runTestsWin("test/unit/math/prim/fun/mdivide_right_tri_test.cpp", false, false)                        
+                                runTestsWin("test/unit/math/rev/fun/mdivide_left_tri_test.cpp", false, false)
+                                runTestsWin("test/unit/math/rev/fun/multiply_test.cpp", false, false)
```
Comment (t4c1) [inline Jenkinsfile:263]: Can you reorder these so that the same order is used for all combinations of windows/linux CPU/GPU?

**Review by t4c1 — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (rok-cesnovar) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 2 above)
Comment (rok-cesnovar) [reply Jenkinsfile]: Done.

**Review by t4c1 — APPROVED** [review]

**Pull request merged**